src/main/java/core/basesyntax/dao/Box.java

```diff
@@ -0,0 +1,91 @@
+package core.basesyntax.dao;
+
+import core.basesyntax.goods.FruitPack;
+import java.time.LocalDate;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class Box {
+    private final Map<LocalDate, FruitPack> box;
+
+    public Box() {
+        box = new HashMap<>();
+    }
+
+    public Box(FruitPack product) {
+        box = new HashMap<>();
+        this.putProduct(product);
+    }
+
+    public Box putProduct(FruitPack product) {
+        FruitPack.isPresent(product);
+        LocalDate expDate = product.getExpDate();
+        FruitPack.checkExpDate(expDate);
+        if (!box.isEmpty() && box.containsKey(expDate)) {
+            int newQuantity = product.getQuantity();
+            int prevQuantity = box.get(expDate).getQuantity();
+            product = product.setQuantity(newQuantity + prevQuantity);
+        }
+        box.put(expDate, product);
+        return this;
+    }
+
+    public Box takeProduct(FruitPack product) {
+        FruitPack.isPresent(product);
+        LocalDate expDate = product.getExpDate();
+        FruitPack.checkExpDate(expDate);
+        if (!box.isEmpty() && box.containsKey(expDate)) {
+            int takingQuantity = product.getQuantity();
+            int existingQuantity = box.get(expDate).getQuantity();
+            if (takingQuantity > existingQuantity) {
+                throw new IllegalArgumentException("Lack of products");
+            } else if (existingQuantity > takingQuantity) {
+                product = product.setQuantity(existingQuantity - takingQuantity);
+                box.put(expDate, product);
+            } else {
+                box.remove(expDate);
+            }
+        }
+        return this;
+    }
+
+    public FruitPack getProduct(LocalDate expDate) {
+        FruitPack.checkExpDate(expDate);
+        return box.get(expDate);
+    }
+
+    public void clear() {
+        box.clear();
+    }
+
+    public int size() {
+        return box.size();
+    }
+
+    public boolean isEmpty() {
+        return box.isEmpty();
+    }
+
```

Comment:
I think all of these methods are useless because you can call all of them already on the map

Reply:
but I can't reach them from the outside of class, because of data structure access modifier

Reply:
Agree 👍